tests/MongoDB.EntityFrameworkCore.FunctionalTests/Design/CompiledModelTests.cs

```diff
@@ -0,0 +1,177 @@
+/* Copyright 2023-present MongoDB Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System.Runtime.CompilerServices;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Design;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Scaffolding;
+using Microsoft.Extensions.DependencyInjection;
+using MongoDB.Bson;
+using MongoDB.EntityFrameworkCore.Design;
+
+namespace MongoDB.EntityFrameworkCore.FunctionalTests.Design;
+
+[XUnitCollection("DesignTests")]
+public class CompiledModelTests(TemporaryDatabaseFixture tempDatabase)
+    : IClassFixture<TemporaryDatabaseFixture>
+{
+    public class EveryType
+    {
+        public ObjectId id { get; set; }
+
+        public Decimal128 aDecimal128 { get; set; }
+
+        public string aString { get; set; }
+        public Guid aGuid { get; set; }
+
+        public int anInt { get; set; }
+        public short aShort { get; set; }
+        public long aLong { get; set; }
+
+        public decimal aDecimal { get; set; }
+        public float aFloat { get; set; }
+        public double aDouble { get; set; }
+
+        public string[] aStringArray { get; set; }
+        public List<int> anIntList { get; set; }
+
+        public List<OwnedEntity> ownedEntities { get; set; }
+    }
+
+    public class OwnedEntity
+    {
+        public string name { get; set; }
+    }
+
+    public class SimpleContext(DbContextOptions options)
+        : DbContext(options)
+    {
+        public DbSet<EveryType> EveryTypes { get; set; }
+    }
+
+    [Fact]
+    public void Can_generate_code_for_model()
+    {
+        var (db, scope) = GetDesignTimeConfigured<SimpleContext>();
+        var files = GenerateModel(db);
+        scope.Dispose();
+
+        Assert.Single(files, f => f.Path == nameof(SimpleContext) + "Model.cs");
+        Assert.Single(files, f => f.Path == nameof(SimpleContext) + "ModelBuilder.cs");
+        Assert.Single(files, f => f.Path == nameof(EveryType) + "EntityType.cs");
+        Assert.Single(files, f => f.Path == nameof(OwnedEntity) + "EntityType.cs");
+    }
+
+    [Fact]
+    public void Regenerate_model_code_for_roundtrip_tests()
+    {
+        // This will re-generate local code-generated model files for SimpleContext
+        // If they change you should run the tests again and check them in.
```

Comment:
Can we try to find more elegant way to re-generate the code? It might be an option to generate the code on fly and load the generated assembly dynamically, so we do not even need to check-in the generated code.

Reply:
I could look into that in the future but I think it's a lot lower pri than a lot of stuff on the backlog right now. I'll raise a ticket and put it in there.

Reply:
EF9 and .NET 9 actually have a build process that does this automatically so I think we should just use that in due course.

Reply:
Given that EF9 solves this exact problem in a standard way, I agree with the approach of filing an EF ticket to track this work in EF9 and checking in the generated test files until that work is completed.